rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/StoragePrinter.scala

```diff
@@ -9,69 +11,87 @@ import coop.rchain.rspace.internal.{Datum, Row, WaitingContinuation}
 import coop.rchain.rspace.trace.{Consume, Produce}
 
 object StoragePrinter {
-  def prettyPrint[F[_]](space: RhoISpace[F]): String = {
-    val pars: Seq[Par] = space.toMap.map {
-      case (
-          channels: Seq[Par],
-          row: Row[BindPattern, ListParWithRandom, TaggedContinuation]
-          ) => {
-        def toSends(data: Seq[Datum[ListParWithRandom]]): Par = {
-          val sends: Seq[Send] = data.flatMap {
-            case Datum(as: ListParWithRandom, persist: Boolean, _: Produce) =>
-              channels.map { channel =>
-                Send(channel, as.pars, persist)
-              }
+  def prettyPrint[F[_]: FlatMap](space: RhoISpace[F]): F[String] =
+    for {
```

Comment:
There is too much code in this for comprehension and it becomes difficult to read. You could move the functions to sends and to receive out of it

Reply:
Done_